Upgrade Hibernate ORM to 6.2.5.Final and Hibernate Reactive to 2.0.1.Final

DavideD opened the pull request:
* Hibernate ORM to 6.2.5.Final
* Hibernate Reactive to 2.0.1.Final

Sorry, I've deleted the previous PR (#34058) by mistake

DavideD committed:
Upgrade Hibernate Reactive and ORM

* Hibernate ORM to 6.2.5.Final
* Hibernate Reactive to 2.0.1.Final

famod commented:
Can this be backported to 3.1.3?

gsmet commented:
@famod for now it cannot be merged due to failures. The upgrade will be backported if we end up being able to merge it.

yrodiere committed:
Include embedded ID types in the types passed to Hibernate ORM for enhancement

yrodiere committed:
Declare additional JPA model elements in relevant extensions

gsmet commented on extensions/smallrye-reactive-messaging-kafka/deployment/pom.xml:
```diff
@@ -42,6 +42,10 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-jackson-deployment</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-orm-deployment</artifactId>
+        </dependency>
```
I don't think you can add a deployment dependency to an extension without also adding the runtime implementation. And that is going to be an issue.

Looks like this build item should be an SPI module (unfortunately).

yrodiere replied:
Ok, but there *are* dependencies to the runtime in the runtime module, they're just optional:

https://github.com/quarkusio/quarkus/blob/ba8c0fef65051e2743800a67f2b92cb5c59fc32c/extensions/smallrye-reactive-messaging-kafka/runtime/pom.xml#L34-L43

I'll move this to SPI, but I wonder if the requirements should be adjusted to take into account optional dependencies.

yrodiere commented:
I confirmed the Spring Data failure as a regression in ORM (see https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev/topic/Bytecode.20enhancement.20and.20embedded.20ID ; I'll open an issue later).

@gsmet Do you want me to find a workaround and merge anyway, or wait for the next version of ORM that fixes this regression?

yrodiere commented:
I filed https://hibernate.atlassian.net/browse/HHH-16832 for the ORM bytecode enhancement bug. I will attempt to workaround the bug in our test suite for now.

yrodiere committed:
Work around https://hibernate.atlassian.net/browse/HHH-16832

yrodiere committed:
Move AdditionalJpaModelBuildItem to an SPI module

So that it may be used by smallrye-reactive-messaging, for whom the
Hibernate ORM extension is just an optional dependency.

yrodiere commented:
Alright. I moved `AdditionalJpaModelBuildItem` to the SPI module to pass the automated check, and worked around https://hibernate.atlassian.net/browse/HHH-16832 in Spring Data JPA tests. The Hibernate ORM team is going to investigate the HHH-16832 in its next sprint.

yrodiere commented:
Test failures are the usual suspects (flaky tests).

@gsmet still ok to merge the workaround?

ChMThiel commented:
when will hibernate 6.2.5 will be available? We need this fix, because of this [hibernate-bug](https://hibernate.atlassian.net/browse/HHH-16612) we are not able to update.

yrodiere commented:
> @gsmet still ok to merge the workaround?
